src/controllers/authentication/session.rs

```diff
@@ -43,43 +38,84 @@ pub async fn login(
-        Identity::login(&req.extensions(), username.to_string())
-            .map_err(actix_web::error::ErrorInternalServerError)?;
-        Ok(HttpResponse::SeeOther()
-            .insert_header((LOCATION, "/admin/posts/page/1"))
-            .finish())
-    } else {
+    // if no user exists then null will be returned from DB
+    // None will be returned in option if the user does not exist
+    // check is no user exits then throw a flash message
+    if parsed_hash.is_none() {
         FlashMessage::error("Login Fail - Wrong Id or password!").send();
+
         Ok(HttpResponse::SeeOther()
             .insert_header((http::header::LOCATION, "/login"))
             .finish())
+    } else {
+        let struct_only = parsed_hash.unwrap_or_default();
+        let PasswordStruct { password } = struct_only;
+        // insert password from DB to password hash
+        let parsed_stored =
+            PasswordHash::new(&password).map_err(actix_web::error::ErrorInternalServerError)?;
+        // check the user password and check the password from database
+        let valid_user = Argon2::default()
+            .verify_password(password_input.as_bytes(), parsed_stored.borrow())
+            .map_err(actix_web::error::ErrorInternalServerError);
+        // check the verify_password is successful
+        // or failure using match
+        match valid_user {
+            Ok(_) => {
+                Identity::login(&req.extensions(), username)
+                    .map_err(actix_web::error::ErrorInternalServerError)?;
+
+                Ok(HttpResponse::SeeOther()
+                    .insert_header((LOCATION, "/admin/posts/page/1"))
+                    .finish())
+            }
+            Err(_) => {
+                FlashMessage::error("Login Fail - Wrong Id or password!").send();
+
+                Ok(HttpResponse::SeeOther()
+                    .insert_header((http::header::LOCATION, "/login"))
+                    .finish())
+            }
+        }
     }
 }
-
 pub async fn logout(id: Identity) -> impl Responder {
     id.logout();
     web::Redirect::to("/")
 }
 
 pub async fn check_user(user: Option<Identity>) -> impl Responder {
+    // check of the user exists by using cookie from browser
     if let Some(_user) = user {
         web::Redirect::to("/admin/posts/page/1")
     } else {
         web::Redirect::to("/")
     }
 }
 pub fn build_message_framework(signing_key: Key) -> FlashMessagesFramework {
+    // send flash messages using cookies
+    // --> CookieMessageStore will send the data to the browser
     let message_store = CookieMessageStore::builder(signing_key).build();
     FlashMessagesFramework::builder(message_store).build()
 }
+
+#[derive(Deserialize, Debug, Clone, PartialEq, sqlx::FromRow)]
+pub struct PasswordStruct {
+    pub password: String,
+}
+
+impl Default for PasswordStruct {
+    fn default() -> Self {
+        PasswordStruct {
+            password: "password".to_string(),
+        }
+    }
+}
```

Comment:
```rust

} else {
        let struct_only = parsed_hash.unwrap_or_default();
        let PasswordStruct { password } = struct_only;
        
        ```

I have to unwrap the user entered password. So I have used unwrap or default .

It might not be the best technique